### PR TITLE
privilege: support authentication with active roles

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve? 
User can use roles' privileges if roles are granted to user and active. 

### What is changed and how it works?
When a user action need to be checked privileges.  `RequestVerification` method  will search all roles associate with that roles, and gather roles privileges using logic or.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Possible performance regression
 - Increased code complexity

### CODE REVIEW

File: planner/core/optimizer.go [hunk 1]
```diff
@@ -14,6 +14,7 @@
 package core
 
 import (
+	"github.com/pingcap/parser/auth"
```
Comment (jackysp) [inline planner/core/optimizer.go]: move it down with the 3rd part libs

File: privilege/privileges/cache.go [hunk 2]
```diff
@@ -580,29 +607,52 @@ func (p *MySQLPrivilege) matchColumns(user, host, db, table, column string) *col
...
-func (p *MySQLPrivilege) RequestVerification(user, host, db, table, column string, priv mysql.PrivilegeType) bool {
-	record1 := p.matchUser(user, host)
-	if record1 != nil && record1.Privileges&priv > 0 {
+func (p *MySQLPrivilege) RequestVerification(activeRoles []*auth.RoleIdentity, user, host, db, table, column string, priv mysql.PrivilegeType) bool {
+	roleList := p.FindAllRole(activeRoles)
+	roleList = append(roleList, &auth.RoleIdentity{Username: user, Hostname: host})
+
+	var userPriv, dbPriv, tablePriv, columnPriv mysql.PrivilegeType
+	for _, r := range roleList {
+		record1 := p.matchUser(r.Username, r.Hostname)
```
Comment (jackysp) [inline privilege/privileges/cache.go]: record1, 2, 3, 4 are hard for the developers to understand.

File: privilege/privileges/cache.go [hunk 3]
```diff
@@ -580,29 +607,52 @@ func (p *MySQLPrivilege) matchColumns(user, host, db, table, column string) *col
 }
 
 // RequestVerification checks whether the user have sufficient privileges to do the operation.
-func (p *MySQLPrivilege) RequestVerification(user, host, db, table, column string, priv mysql.PrivilegeType) bool {
-	record1 := p.matchUser(user, host)
-	if record1 != nil && record1.Privileges&priv > 0 {
+func (p *MySQLPrivilege) RequestVerification(activeRoles []*auth.RoleIdentity, user, host, db, table, column string, priv mysql.PrivilegeType) bool {
```
Comment (tiancaiamao) [inline privilege/privileges/cache.go]: The more things a function does, the less easier to combinate the function.
Could we keep the original interface, and wrap it to provide the new function? 

```
func RequestVerification(p *MySQLPrivilege, activeRoles []*auth.RoleIdentity, ...) {
    for _, role := range activeRoles {
        p.RequestVerification(p, role.user, role.host, ...)
    }
}
```

File: privilege/privileges/privileges.go [hunk 4]
```diff
@@ -76,7 +76,8 @@ func (p *UserPrivileges) RequestVerificationWithUser(db, table, column string, p
 	}
 
 	mysqlPriv := p.Handle.Get()
-	return mysqlPriv.RequestVerification(user.Username, user.Hostname, db, table, column, priv)
+	tmpRoleList := make([]*auth.RoleIdentity, 0)
+	return mysqlPriv.RequestVerification(tmpRoleList, user.Username, user.Hostname, db, table, column, priv)
```
Comment (tiancaiamao) [inline privilege/privileges/privileges.go]: How about mysqlPriv.RequestVerification(**nil**, ...)

**Comment by tiancaiamao** [comment]
LGTM
PTAL @jackysp

**Comment by tiancaiamao** [comment]
PTAL @jackysp

**Pull request merged**